libc/string_catdup.c

```diff
@@ -0,0 +1,689 @@
+	TEST_ASSERT_EQUAL_STRING(WORD_1 WORD_2, buf1);
+
+	free(buf1);
+}
+
+
+TEST(string_dup, strdup_empty)
+{
+	char buf[SIZE_BUF] = "";
+	char *buf1 = strdup(buf);
+
+	TEST_ASSERT_NOT_NULL(buf1);
+
+	TEST_ASSERT_EQUAL_STRING(buf, buf1);
+	TEST_ASSERT_EQUAL_STRING("", buf1);
+
+	free(buf1);
+}
+
+
+TEST(string_dup, strdup_huge_string)
+{
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wstringop-truncation"
+	char huge_buf[PATH_MAX] = "";
+	char buf[PATH_MAX] = "";
+
+	for (int i = 0; i < PATH_MAX - 1; i++) {
+		huge_buf[i] = HUGE_STRING[i];
+		buf[i] = HUGE_STRING[i];
+	}
+
+	huge_buf[PATH_MAX - 1] = '\0';
+	buf[PATH_MAX - 1] = '\0';
+
+	char *buf1 = strdup(buf);
+
+	TEST_ASSERT_NOT_NULL(buf);
+	TEST_ASSERT_NOT_NULL(buf1);
+
+	TEST_ASSERT_EQUAL_STRING(buf, buf1);
+	TEST_ASSERT_EQUAL_STRING(huge_buf, buf1);
+
+	free(buf1);
+#pragma GCC diagnostic pop
+}
+
+
+TEST(string_dup, strdup_special_char)
+{
+	char *buf = SPECIAL_CHAR;
+	char *buf1 = strdup(buf);
+
+	TEST_ASSERT_NOT_NULL(buf1);
+
+	TEST_ASSERT_EQUAL_STRING(buf, buf1);
+	TEST_ASSERT_EQUAL_STRING(SPECIAL_CHAR, buf1);
+
+	free(buf1);
+}
+
+
+TEST(string_cat, strdup_ascii)
+{
+	int c, count = 0;
+	char ascii[256] = { 0 };
+
+	for (c = 0; c < 255; c++) {
+		if (isprint(c)) {
+			ascii[count++] = c;
+		}
+	}
+	ascii[count++] = '\0';
+
+	char *buf = strdup(ascii);
+
+	TEST_ASSERT_EQUAL_STRING(ascii, buf);
+
+	free(buf);
+}
+
+
+TEST(string_dup, strdup_non_ascii)
+{
+	char *buf = NON_ASCII_1 NON_ASCII_2;
+	char *buf1 = strdup(buf);
+
+	TEST_ASSERT_NOT_NULL(buf1);
+
+	TEST_ASSERT_EQUAL_STRING(buf, buf1);
+	TEST_ASSERT_EQUAL_STRING(NON_ASCII_1 NON_ASCII_2, buf1);
+
+	free(buf1);
+}
+
+
+TEST(string_dup, strndup_basic)
+{
+	char buf[SIZE_BUF] = WORD_1 WORD_2;
+	char *buf1 = strndup(buf, 5);
+
+	TEST_ASSERT_NOT_NULL(buf1);
+
+	TEST_ASSERT_EQUAL_STRING(WORD_1 WORD_2, buf);
+	TEST_ASSERT_EQUAL_STRING("Lorem", buf1);
+
+	free(buf1);
+}
+
+
+TEST(string_dup, strndup_full_string)
+{
+	char buf[SIZE_BUF] = WORD_1 WORD_2;
+	char *buf1 = strndup(buf, strlen(buf));
+
+	TEST_ASSERT_NOT_NULL(buf1);
+
+	TEST_ASSERT_EQUAL_STRING(buf, buf1);
+	TEST_ASSERT_EQUAL_STRING(WORD_1 WORD_2, buf1);
+
+	free(buf1);
+}
+
+
+TEST(string_dup, strndup_zero_size)
+{
+	char buf[SIZE_BUF] = WORD_1 WORD_2;
+	char *buf1 = strndup(buf, 0);
+
+	TEST_ASSERT_NOT_NULL(buf1);
+
+	TEST_ASSERT_EQUAL_STRING(WORD_1 WORD_2, buf);
+	TEST_ASSERT_EQUAL_STRING("", buf1);
+
+	free(buf1);
+}
+
+
+TEST(string_dup, strndup_large_size)
+{
+#ifndef __phoenix__
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wstringop-overread"
+#endif
+	errno = 0;
+	char buf[SIZE_BUF] = WORD_1 WORD_2;
+	char *buf1 = strndup(buf, SIZE_BUF);
+
+	TEST_ASSERT_NOT_NULL(buf1);
+	TEST_ASSERT_EQUAL_INT(0, errno);
+
+	TEST_ASSERT_EQUAL_STRING(buf, buf1);
+	TEST_ASSERT_EQUAL_STRING(WORD_1 WORD_2, buf1);
+
+	free(buf1);
+#ifndef __phoenix__
+#pragma GCC diagnostic pop
+#endif
+}
+
+
+TEST(string_dup, strndup_special_char)
```

Comment:
```suggestion
TEST(string_dup, strndup_specials)
```

or alternatively `special_chars`